Allow compiling the `wasm32-wasi` std library with atomics

abrown opened the pull request:
The issue #102157 demonstrates how currently the `-Z build-std` option will fail when re-compiling the standard library with `RUSTFLAGS` like `RUSTFLAGS="-C target-feature=+atomics,+bulk-memory -C link-args=--shared-memory"`. This change attempts to resolve those build issues by depending on the the WebAssembly `futex` module and providing an implementation for `env_lock`. Fixes #102157.

abrown committed:
Allow compiling the `wasm32-wasi` std library with atomics

The issue #102157 demonstrates how currently the `-Z build-std` option
will fail when re-compiling the standard library with `RUSTFLAGS` like
`RUSTFLAGS="-C target-feature=+atomics,+bulk-memory -C
link-args=--shared-memory"`. This change attempts to resolve those build
issues by depending on the the WebAssembly `futex` module and providing
an implementation for `env_lock`. Fixes #102157.

rust-highfive commented:
r? @thomcc

(rust-highfive has picked a reviewer for you, use r? to override)

joboet commented:
>     1. I assume that `env_lock` here should actually return one of the lock guards available to WebAssembly but I'm not really sure yet about the details of this — can someone point me to an example of how this should be done from other targets?

[UNIX uses an `RwLock`](https://github.com/rust-lang/rust/blob/57ee5cf5a93923dae9c98bffb11545fc3a31368d/library/std/src/sys/unix/os.rs#L506) to ensure correct synchronization between setting and getting the environment variables, as `setenv` is unsynchronized. The `libc` used on WASI is based on MUSL and [is not thread-safe](https://github.com/WebAssembly/wasi-libc/blob/main/libc-top-half/musl/src/env/putenv.c), too, so it's probably best to copy the UNIX implementation.

alexcrichton commented:
> But I'm not sure how to correctly use this new sysroot to compile a Rust program; a command like the following fails — what is the right way to do this?

You'll need to set `lld = true` in `config.toml` as well to ensure that you're locally compiling `wasm-ld`. Otherwise though I don't think WASI tests work with the test harness right now and I don't think that `compiletest` has support for WASI. Notably [this](https://github.com/rust-lang/rust/blob/master/src/etc/wasm32-shim.js) has no support for WASI which is used as a node.js-based runner for wasm tests.

I've historically dabbled in getting `wasm32-wasi` tests up and running using `wasmtime`-the-CLI but I never ended up finishing the work or proposing it here since it didn't feel like it fit. Overall for testing this I think the best thing to do for now is to compile a toolchain and test it on local programs. Eventually when the whole test suite is up and running more issues can be ironed out.

(FWIW historical test runs turned up no actual failures, just a lot of infrastructural changes to get things working)

alexcrichton commented:
If you're interested [this is my historical work](https://github.com/alexcrichton/rust/commit/63edbe6826bb73c886169c65605e6619fc22410d) to get the test suite running with `wasmtime`-the-CLI. requiring you to specify:

```toml
[target.wasm32-wasi]
runtool = 'wasmtime run --disable-cache --'
```

in `config.toml`. This doesn't pass the test suite since lots of `// ignore-...` needs updates for `wasm32-wasi`, but you should be able to skim the failures and see if there's anything obvious. I don't think that the `src/test/ui` test suite will be all that useful for testing this change though.

thomcc commented:
This is still a draft. Feel free to kick it back to me with at-rustbot ready once it's ready for review.

@rustbot author

abrown committed:
Implement `env_lock` with `RwLock`

Copying the approach of the Unix target, this change uses the standard
`RwLock` to protect against concurrent access of libc's environment.
This locking is only enabled when WebAssembly's `atomics` feature is
also enabled.

joboet commented on library/std/src/sys/wasi/os.rs:
```diff
@@ -23,13 +23,24 @@ mod libc {
+    if #[cfg(target_feature = "atomics")] {
+        // Access to the environment must be protected by a lock in multi-threaded scenarios.
+        use crate::sys_common::rwlock::StaticRwLock;
+        static ENV_LOCK: StaticRwLock = StaticRwLock::new();
```
`StaticRwLock` has been removed (see #100581), use `sync::RwLock` instead (consequently these functions don't need to be `unsafe` anymore).

abrown commented:
As @alexcrichton notes above, the in-tree testing story still has some way to go, so I manually tested that this change works using the following sequence of steps:

1. First, I built the latest HEAD of [`wasi-libc`](https://github.com/WebAssembly/wasi-libc) with threads enabled:

  ```console
  $ git rev-parse HEAD
  05b3b876e63cdc919c79f5127a7ae8dbdb7b095f
  $ make THREAD_MODEL=posix
  ```

2. Then, in this repository, I altered my `config.toml` to use this new sysroot:

  ```console
  // first, enabled building the linker by setting `lld = true` in the [rust] section; then:
  $ echo '[target.wasm32-wasi]' >> config.toml
  $ echo 'wasi-root = "/.../wasi-libc/sysroot"' >> config.toml
  ```

3. Then, I built the Rust toolchain with all of the changes in this PR along with the flags to activate them:

  ```console
  $ RUSTFLAGS="-C target-feature=+atomics,+bulk-memory" ./x.py build library --target wasm32-wasi
  $ rustup toolchain link stage1 build/x86_64-unknown-linux-gnu/stage1
  ```

5. Using the same [`kernel.rs`](https://github.com/rust-lang/rust/issues/102157#issue-1383047794) example from #102157, I was able to successfully compile a Rust program that uses shared memory:

  ```console
  $ rustc +stage1 kernel.rs -o kernel.wasm --target wasm32-wasi --crate-type=cdylib  -C link-args="--shared-memory" -C target-feature='+atomics,+atomics'
  $ wasm2wat --enable-threads kernel.wasm | grep '(memory'
    (import "env" "memory" (memory (;0;) 17 16384 shared))
  ```

abrown commented:
I rebased this change on latest `master` and switched to using `RwLock` as @joboet pointed out. If someone can point me to the right way to encode a test based on the manual steps I did above, I will add that, but otherwise I think this is ready for review.

@rustbot ready

abrown commented:
Ok, I guess: @rustbot label +T-libs-api -T-libs

abrown commented:
Oh, never mind: @rustbot label -T-libs-api +T-libs

thomcc commented:
Have you done a check for the places we might `cfg(target_arch = "wasm")` to see if any of them have similar assumptions? My main hesitation here is I seem to recall that there's a fair bit of code that makes the same assumption (possibly including code `std` has a dependency on).

abrown commented:
> Have you done a check for the places we might `cfg(target_arch = "wasm")` to see if any of them have similar assumptions? My main hesitation here is I seem to recall that there's a fair bit of code that makes the same assumption (possibly including code `std` has a dependency on).

I think you mean `target_family = "wasm"`? Once I look for that I see that `library/std/src/thread/local.rs` has quite a few references to that but it seems to me that they are appropriately paired with `not(target_feature = "atomics")`. Perhaps that was what you were concerned about? The entire repository has 26 results for `target_family = "wasm"` but the rest do not seem to me to pertain to your question.

I would say that compiling Rust with atomics and shared memory is still a highly risky venture and that this PR only fixes an issue that I am aware of. There will likely be others once people start experimenting with this.

thomcc commented:
> I think you mean target_family = "wasm"

Uh, yeah, and just a grep for cfgs with wasm in them I guess (in case they were on `target_arch="wasm32"` or something).

> I would say that compiling Rust with atomics and shared memory is still a highly risky venture and that this PR only fixes an issue that I am aware of. There will likely be others once people start experimenting with this.

This is pretty concerning (assuming you mean that there are likely issues in `std`; I'm aware the ecosystem is... somwhat ill-prepared for this). Is this target_feature stable? If so I think it might be better to get a handle on it before we expose users to a bunch of stdlib code that may have bugs...

thomcc commented:
Oh, it requires -Zbuild-std? Hm...

abrown commented:
Yeah, this PR just fixes compilation with features that are already exposed (even if experimentally behind `-Z`). Is there anything else I should do to get this reviewed and merged?

thomcc commented:
I think one option would be to make it clear that it's unstable, by having `target_feature = "atomics"` punt us into [restricted-std](https://github.com/rust-lang/rust/blob/0ca356586fed56002b10920fd21ddf6fb12de797/library/std/build.rs#L37-L48) mode. This would make `std` be considered [unstable](https://github.com/rust-lang/rust/blob/0ca356586fed56002b10920fd21ddf6fb12de797/library/std/src/lib.rs#L192), and require `#![feature(restricted_std)]` to use on wasm+atomics. The benefit of this is that people would understand that it isn't a case of "the target just doesn't ship with a `libstd` so you need to build it" and is really "using std on this target is not stable.

This may be overkill though, or the wrong approach entirely. It makes me very uncomfortable to allow this to compile (even with all this) to consider allowing this to build before someone has gone through and audited `std` (and the dependencies of `std` that are used on wasm platforms) for places where we assume that wasm has no threads (and don't bother bother with locks/TLS/some other dangerous outcome).

I've nominated this for discussion at next week's libs meeting, and will try to do all or part of that audit before then.

abrown commented:
> (Like, are we really comfortable, even on nightly, letting Rust users use a std which just might be wildly thread unsafe?)

Well, considering that no one was using this `std`, since it wouldn't compile until this PR, I wasn't too concerned about the possibility of many users seeing issues. I will admit that I'm not too familiar with the restrictions you are talking about, so if you are convinced that things should/could be safer, I'm interested in whatever you suggest (especially if you do the audit, since I'm not exactly clear how "wild" things need to be to be "wildly unsafe").

thomcc commented:
> unless anything major has changed in the meantime it should still be the case that for wasm32-unknown-unknown things should work.
> ...
> I'm also more than happy to vouch for both the correctness of this patch and helping to shepherd along any threads-related issues in the wasm target.

Both of these are very good to hear, and pretty much alleviate my concerns.

> The threads proposal itself to WebAssembly has not "officially shipped" and additionally toolchains have not at this time generally figured out how to get traction. The target support in Rust is largely "you know everything about what you're doing and you're trying to get something working"...

So, on the one hand, I had understood that this *is* the kind of situation that restricted_std is for. But I'm sympathetic to the fact that it could cause a ton of ecosystem churn (both now to add the `feature(restricted_std)` under wasm+threads and then later to remove it, when the target becomes ready for broader use), and that there would be no way to experiment around this stuff in the meantime. I think this means that perhaps `restricted_std` isn't really a great way to do things!

Given that it *does* already require nightly to compile (due to needing `-Zbuild-std`), I probably am inclined to agree that this is overkill, but we should be somewhat careful before shipping a build that users can actually use.

I'll leave the nomination for now, but I expect we'll land this. FWIW, If you'd like to attend the meeting, you're more than welcome -- we don't have video meetings for T-libs anymore (only t-libs-api currently has those), so it's just a matter of being on zulip around 8AM PT on Weds in the [t-libs/meetings](https://rust-lang.zulipchat.com/#narrow/stream/259402-t-libs.2Fmeetings) zulip stream. (But no worries if you can't -- I think the discussion perhaps should be more about when we should/should not be using `restricted_std`)

bors commented:
:pushpin: Commit 587e971beaebb3903c64d25679d062dc6079844d has been approved by `thomcc`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=thomcc 587e971beaebb3903c64d25679d062dc6079844d -->
<!-- homu: {"type":"Approved","sha":"587e971beaebb3903c64d25679d062dc6079844d","approver":"thomcc","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #93668) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

abrown commented:
Ok, rebased.

@rustbot label -S-waiting-on-author +S-waiting-on-review

alexcrichton commented:
@thomcc I'm happy to join in discussions and I should be online at that time, so feel free to ping me if I can help out.

thomcc commented:
@bors r+

bors commented:
:pushpin: Commit 9530ba0fe2118d9a60e16861faad29ec5803d3f9 has been approved by `thomcc`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=thomcc 9530ba0fe2118d9a60e16861faad29ec5803d3f9 -->
<!-- homu: {"type":"Approved","sha":"9530ba0fe2118d9a60e16861faad29ec5803d3f9","approver":"thomcc","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

JohnTitor commented:
Failed on rollup: https://github.com/rust-lang/rust/pull/102908#issuecomment-1274273051
@bors r-

abrown committed:
fix: return type of single-threaded dummy lock must be droppable

abrown commented:
@thomcc, 95b0b2d should fix the bors failure.

@rustbot label -S-waiting-on-author +S-waiting-on-review

thomcc commented:
@bors r+

bors commented:
:pushpin: Commit 95b0b2d34956c232ace714c1f5971b20e2767c6c has been approved by `thomcc`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=thomcc 95b0b2d34956c232ace714c1f5971b20e2767c6c -->
<!-- homu: {"type":"Approved","sha":"95b0b2d34956c232ace714c1f5971b20e2767c6c","approver":"thomcc","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 95b0b2d34956c232ace714c1f5971b20e2767c6c with merge 2a9217601c0e5ae83fa8bca6797137a6d2f7e4e5...
<!-- homu: {"type":"BuildStarted","head_sha":"95b0b2d34956c232ace714c1f5971b20e2767c6c","merge_sha":"2a9217601c0e5ae83fa8bca6797137a6d2f7e4e5"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3238858894/jobs/5308918100)
Approved by: thomcc
Pushing 2a9217601c0e5ae83fa8bca6797137a6d2f7e4e5 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"thomcc","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3238858894/jobs/5308918100"},"merge_sha":"2a9217601c0e5ae83fa8bca6797137a6d2f7e4e5"} -->

rust-timer commented:
Finished benchmarking commit (2a9217601c0e5ae83fa8bca6797137a6d2f7e4e5): [comparison URL](https://perf.rust-lang.org/compare.html?start=0938e1680daf66ca6aad428aedf9a920a0dab5ad&end=2a9217601c0e5ae83fa8bca6797137a6d2f7e4e5&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=0938e1680daf66ca6aad428aedf9a920a0dab5ad&end=2a9217601c0e5ae83fa8bca6797137a6d2f7e4e5&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 0.8%     | [0.7%, 1.0%] | 6         |
| Regressions ❌ <br /> (secondary) | -        | -     | 0         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | 0.8%     | [0.7%, 1.0%] | 6         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=0938e1680daf66ca6aad428aedf9a920a0dab5ad&end=2a9217601c0e5ae83fa8bca6797137a6d2f7e4e5&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 2.6%     | [2.6%, 2.6%] | 1         |
| Regressions ❌ <br /> (secondary) | -        | -     | 0         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | 2.6%     | [2.6%, 2.6%] | 1         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=0938e1680daf66ca6aad428aedf9a920a0dab5ad&end=2a9217601c0e5ae83fa8bca6797137a6d2f7e4e5&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 1.9%     | [1.9%, 1.9%] | 1         |
| Regressions ❌ <br /> (secondary) | 3.5%     | [3.5%, 3.5%] | 1         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | 1.9%     | [1.9%, 1.9%] | 1         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

lqd commented:
The regressions are recent diesel noise, and unrelated to this PR and the wasi target.

@rustbot label: +perf-regression-triaged
